testing.rst

```diff
@@ -11,377 +11,513 @@ using both functional and unit tests.
+.. configuration-block::
 
-    # run all tests in the Util/ directory
-    $ php bin/phpunit tests/Util
+    .. code-block:: yaml
 
-    # run tests for the Calculator class
-    $ php bin/phpunit tests/Util/CalculatorTest.php
+        # config/packages/test/twig.yaml
+        twig:
+            strict_variables: true
 
-.. index::
-   single: Tests; Functional tests
+    .. code-block:: xml
 
-.. _functional-tests:
+        <!-- config/packages/test/twig.xml -->
+        <?xml version="1.0" encoding="UTF-8" ?>
+        <container xmlns="http://symfony.com/schema/dic/services"
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+            xmlns:twig="http://symfony.com/schema/dic/twig"
+            xsi:schemaLocation="http://symfony.com/schema/dic/services
+                https://symfony.com/schema/dic/services/services-1.0.xsd
+                http://symfony.com/schema/dic/twig
+                https://symfony.com/schema/dic/twig/twig-1.0.xsd">
 
-Functional Tests
-----------------
+            <framework:config strict-variables="true"/>
+        </container>
 
-Functional tests check the integration of the different layers of an
-application (from the routing to the views). They are no different from unit
-tests as far as PHPUnit is concerned, but they have a very specific workflow:
+    .. code-block:: php
 
-* Make a request;
-* Click on a link or submit a form;
-* Test the response;
-* Rinse and repeat.
+        // config/packages/test/twig.php
+        $container->loadFromExtension('twig', [
+            'strict_variables' => true,
+        ]);
 
-Before creating your first test, install these packages that provide some of the
-utilities used in the functional tests:
+You can also use a different environment entirely, or override the default
+debug mode (``true``) by passing each as options to the ``bootKernel()``
+method::
 
-.. code-block:: terminal
+    self::bootKernel([
+        'environment' => 'my_test_env',
+        'debug'       => false,
+    ]);
 
-    $ composer require --dev symfony/browser-kit symfony/css-selector
+.. tip::
 
-Your First Functional Test
-~~~~~~~~~~~~~~~~~~~~~~~~~~
-Functional tests are PHP files that typically live in the ``tests/Controller``
-directory of your application. If you want to test the pages handled by your
-``PostController`` class, start by creating a new ``PostControllerTest.php``
-file that extends a special ``WebTestCase`` class.
+    It is recommended to run your test with ``debug`` set to ``false`` on
+    your CI server, as it significantly improves test performance.
 
-As an example, a test could look like this::
+Customizing Environment Variables
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-    // tests/Controller/PostControllerTest.php
-    namespace App\Tests\Controller;
+If you need to customize some environment variables for your tests (e.g. the
+``DATABASE_URL`` used by Doctrine), you can do that by overriding anything you
+need in your ``.env.test`` file:
```

Comment:
Environment variables can also be added to your `phpunit.xml`. When would you use `phpunit.xml` over `.env.test`? Is there a reason [these two env variables](https://github.com/symfony/recipes/blob/master/symfony/phpunit-bridge/5.3/phpunit.xml.dist#L14-L15) can't be added to `.env.test`?

Reply:
I'm not sure, I think they can be part of `.env.test`. @nicolas-grekas do you maybe know why they aren't?